Behavioral Guided Single-Cell Transcriptomics Analysis (Part 2)
Different T Cell Subset Identification and Annotation

Note: Before running this analysis workflow, either make sure you have already run the previous part (part1) of the analysis workflow, or you have downloaded corresponding tables from Zenodo depository and have them in the current working directory.

Warning: Based on different factors such as your operating system, R version, Seurat and other package versions the analysis results of different sections can be slightly different. However, major conclusions should not change. If you want to reproduce downstream analysis exactly same, please use the provided Seurat objects in Zenodo depository.

This workflow aims to identify different T cell subsets in the dataset and annotate them. For this purpose, we will conduct a three step analysis workflow:

Read-in the data and remove low quality cells.

1.1. Load Required Libraries
1.2. Read in the pre-processed Seurat object
1.3. Remove Cycling Cells
1.4. Generate a new Seurat object and remove low quality cells
1.5. Normalize & Scale & Reduce dimentionality of the data
Subset the dataset into two major T cell type (CD8 T and CD4 T) and perform standard single-cell analysis workflow for each subtype.

2.1 Analyze CD8+ T cells
2.1.1 Generate Seurat Object and Embed Cells
2.1.2 Batch Correction
2.1.3 Save Seurat Object for Downstream Analysis
2.2 Analyze CD4+ T cells
2.2.1 Generate Seurat Object and Embed Cells
2.2.2 Further Subset CD4+ T Cells
2.2.3 Analyze CD4+/IL17RB- T Cells
2.2.3.1 Generate Seurat Object and Embed Cells
2.2.3.2 Batch Correction
2.2.3.3 Save Seurat Object for Downstream Analysis
2.2.4 Analyze CD4+/IL17RB+ T Cells
2.2.4.1 Generate Seurat Object and Embed Cells
2.2.4.2 Save Seurat Object for Downstream Analysis
Generate a Seurat Object of all T Cells.

3.1 Generate the Object and Subset to Have Only Verified Cells
3.2 Normalize & Scale & Reduce dimentionality of the data
3.3 Batch Correction
3.4 Save Seurat Object
Before you start, make sure that you have the data in the current working directory.
To set the working directory use the following code:

## 1. Read in the sequencing data and remove low quality cells


### 1.1 Load Required Libraries
For the following workflow, we will need Seurat, dplyr, ggplot2, and SeuratWrappers libraries.
Make sure you have them installed before starting the analysis.
Use the following code to install the required packages:


```{r}
#install.packages(c("Seurat", "dplyr", "ggplot2", "remotes")) # apply this lines if you are using a different enviroment 
#remotes::install_github('satijalab/seurat-wrappers')

suppressMessages(library(Seurat))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(SeuratWrappers))

home <- "path to your folder" # select home directory

setwd(home)
```

### 1.2 Read in the pre-processed ERCC-Removed Seurat object
The input is the set of (per-plate) seurat objects. These objects exclude the 'ERCC genes' that served only technical QC purposes.  

```{r}
plates = c('TM621', 'TM623', 'TM624', 'TM625', 'TM626', 'TM627', 'TM628', 'TM629', 
            'TM632', 'TM713', 'TM714', 'TM715', 'TM716', 'TM721', 'TM722')

# Grab the pre-processed plates, ERCC-genes are removed from the matrices
for (plate in plates) {
  file = paste(plate, "_noE.tab", sep = "") 
  plate_in = read.delim(paste("./2_noERCC/", file, sep = ""), header = TRUE, row.names = 1)
    assign(plate, CreateSeuratObject(counts = plate_in, min.cells = 2, min.features = 200))
}
```

The experimental conditions and (per-plate) annotations are added to the metadata.

```{r}
TM621@meta.data$Engagement = "engager"; TM623@meta.data$Engagement = "nonengager"; TM624@meta.data$Engagement = "super-engaged"
TM625@meta.data$Engagement = "super-engaged"; TM626@meta.data$Engagement = "never-engaged"; TM627@meta.data$Engagement = "never-engaged"
TM628@meta.data$Engagement = "medium-exposed"; TM629@meta.data$Engagement = "medium-exposed"; TM632@meta.data$Engagement = "engager"
TM713@meta.data$Engagement = "super-engaged"; TM714@meta.data$Engagement = "super-engaged"; TM715@meta.data$Engagement = "never-engaged"
TM716@meta.data$Engagement = "never-engaged"; TM721@meta.data$Engagement = "medium-exposed"; TM722@meta.data$Engagement = "medium-exposed"

TM621@meta.data$Organoid = "13T"; TM623@meta.data$Organoid = "13T"; TM624@meta.data$Organoid = "13T"
TM625@meta.data$Organoid = "13T"; TM626@meta.data$Organoid = "13T"; TM627@meta.data$Organoid = "13T"
TM628@meta.data$Organoid = "none"; TM629@meta.data$Organoid = "none"; TM632@meta.data$Organoid = "13T"
TM713@meta.data$Organoid = "13T"; TM714@meta.data$Organoid = "13T"; TM715@meta.data$Organoid = "13T"
TM716@meta.data$Organoid = "13T"; TM721@meta.data$Organoid = "none"; TM722@meta.data$Organoid = "none"

TM621@meta.data$Plate = "TM621"; TM623@meta.data$Plate = "TM623"; TM624@meta.data$Plate = "TM624"
TM625@meta.data$Plate = "TM625"; TM626@meta.data$Plate = "TM626"; TM627@meta.data$Plate = "TM627"
TM628@meta.data$Plate = "TM628"; TM629@meta.data$Plate = "TM629"; TM632@meta.data$Plate = "TM632"
TM713@meta.data$Plate = "TM713"; TM714@meta.data$Plate = "TM714"; TM715@meta.data$Plate = "TM715"
TM716@meta.data$Plate = "TM716"; TM721@meta.data$Plate = "TM721"; TM722@meta.data$Plate = "TM722"

TM621@meta.data$Experiment = "series_2"; TM623@meta.data$Experiment = "series_2"; TM624@meta.data$Experiment = "series_2"
TM625@meta.data$Experiment = "series_2"; TM626@meta.data$Experiment = "series_2"; TM627@meta.data$Experiment = "series_2"
TM628@meta.data$Experiment = "series_2"; TM629@meta.data$Experiment = "series_2"; TM632@meta.data$Experiment = "series_2"
TM713@meta.data$Experiment = "series_3"; TM714@meta.data$Experiment = "series_3"; TM715@meta.data$Experiment = "series_3"
TM716@meta.data$Experiment = "series_3"; TM721@meta.data$Experiment = "series_3"; TM722@meta.data$Experiment = "series_3"
```

Once the (per-plate) objects are properly annotated, they are merged into a single seurat object. 

```{r}
teg_orig = suppressWarnings(merge(x=TM621, y=list(TM623,TM632,TM628, TM629,TM624,TM625,TM626,TM627,TM721,TM722,TM713, 
                                                      TM714,TM715,TM716)))
```

First we look at this merged dataset without any additional correction. In order to have a Normalize & Scale & Reduce dimentionality of the data
For this step we utilized the Guided Clustering Tutorialworkflow, suggested by the authors of Seurat package.

```{r}
teg_orig = NormalizeData(teg_orig, verbose=F)%>% FindVariableFeatures(nfeatures = 4000, verbose=F) %>% ScaleData(features=rownames(teg_orig), verbose=F) %>% RunPCA(verbose=F)

teg_orig = suppressWarnings(RunUMAP(teg_orig, dims=1:30, min.dist=0.2, verbose=F)) %>% FindNeighbors(dims=1:30, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot Cells
DimPlot(teg_orig, group.by="Experiment")
```

The UMAP plot above reflects an experiment related batch effect. MNN batch correction is applied to remove this experimental batch effect.

```{r}
teg_orig = RunFastMNN(object.list=SplitObject(teg_orig, split.by="Experiment"), verbose=F)

teg_orig = RunUMAP(teg_orig, reduction = "mnn", dims = 1:30, verbose=F) %>% FindNeighbors(reduction="mnn", dims=1:30, verbose=F) %>% FindClusters(resolution=0.7, verbose=F)
```

```{r}
options(repr.plot.width=11, repr.plot.height=5)
DimPlot(teg_orig, label=T) + DimPlot(teg_orig, group.by = "Experiment")
```

### 1.3 Remove Cycling Cells
Using known markers (such as CENPA, TOP2A) we identified cycling (proliferating) cells and removed them from the dataset.

```{r}
options(repr.plot.width=11, repr.plot.height=5)
FeaturePlot(teg_orig, order=T, max.cutoff = "q90", min.cutoff = "q10", c("CENPA","TOP2A")) # Cluster 9 is cycling cells.
teg_orig = subset(teg_orig, idents = "9", invert = T)
```

### 1.4 Generate a new Seurat object and remove low quality cells
After generating a new Seurat object we calculate the percentage of mitochondrial and ribosomal genes in each cell and remove cells with high percentage.
Also, cells with exceedingly high or low amount of RNA and genes expressed were removed.

```{r}
# Generate a new Seurat object
teg_counts = GetAssayData(teg_orig, slot = "counts")
teg = CreateSeuratObject(counts = teg_counts, project = "teg", min.cells = 6, min.features = 6)
teg$Engagement = teg_orig$Engagement
teg$Organoid = teg_orig$Organoid
teg$Experiment = teg_orig$Experiment
teg$Plate = teg_orig$Plate
teg[["percent.mt"]] = PercentageFeatureSet(teg, pattern = "MT-")
teg[["percent.rb"]] = PercentageFeatureSet(teg, pattern = "^RP[LS]")
VlnPlot(teg, features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "percent.rb"), ncol = 4)
```

```{r}
# Remove low quality cells
teg = subset(teg, subset = nFeature_RNA > 650 & nFeature_RNA < 4500 & nCount_RNA < 17500 & percent.mt < 15 & percent.rb < 30)

# Remove mito and rb genes (This step is optional, but will reduce the amount of noise during the downstream steps.)
mt.genes = rownames(teg)[grep("^MT-", rownames(teg))]
rb.genes = rownames(teg)[grep("^RP[LS]", rownames(teg))]
genes = c(mt.genes, rb.genes)
genes.keep = setdiff(rownames(teg), genes)
teg = subset(teg, features = genes.keep)
rm(genes, genes.keep, mt.genes, rb.genes)
```

### 1.5 Normalize & Scale & Reduce dimentionality of the data
For this step we utilized the Guided Clustering Tutorialworkflow, suggested by the authors of Seurat package.

```{r}
teg = NormalizeData(teg, verbose=F)%>% FindVariableFeatures(nfeatures = 3000, verbose=F) %>% ScaleData(features=rownames(teg), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
teg = suppressWarnings(RunUMAP(teg, dims=1:30, min.dist=0.2, verbose=F)) %>% FindNeighbors(dims=1:30, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot Cells
options(repr.plot.width=6, repr.plot.height=5)
DimPlot(teg)
```

 

 

## 2. Subset the dataset into two major T cell type (CD8 T and CD4 T)
Based on the expression levels of two marker genes (CD4 and CD8) we subset the dataset into two major T cell subsets.

```{r}
pop1 = WhichCells(teg, expression = CD8A > CD4)
cd8_temp = subset(teg, cells = pop1)
cd8_temp$Engagement = teg$Engagement; cd8_temp$Organoid = teg$Organoid
cd8_temp$Experiment = teg$Experiment; cd8_temp$Plate = teg$Plate
cd4_temp = subset(teg, cells = pop1, invert = T)
cd4_temp$Engagement = teg$Engagement; cd4_temp$Organoid = teg$Organoid
cd4_temp$Experiment = teg$Experiment; cd4_temp$Plate = teg$Plate
```

 

### 2.1 Analyze CD8+ T cells
#### 2.1.1 Generate Seurat Object and Embed Cells
Just like before, we generate a new Seurat object of only CD8+ T cells and analyze them using the Guided Clustering Tutorial workflow.

```{r}
cd8_counts = GetAssayData(cd8_temp, slot="counts")
cd8 = CreateSeuratObject(counts=cd8_counts, project="cd8", min.cells=0, min.features=0)
cd8$Engagement = factor(cd8_temp$Engagement, levels = c("medium-exposed", "never-engaged", "nonengager", "engager", "super-engaged"))
cd8$Engagement = gsub("nonengager", "never-engaged", cd8$Engagement)
cd8$Organoid = cd8_temp$Organoid
cd8$Experiment = cd8_temp$Experiment
cd8$Plate = cd8_temp$Plate

# Normalized & Scale & PCA dimentionality reduction
cd8 = NormalizeData(cd8, verbose=F) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(features=rownames(cd8), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
cd8 = RunUMAP(cd8, dims=1:12, min.dist=0.2, verbose=F) %>% FindNeighbors(dims=1:12, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd8) + DimPlot(cd8, group.by="Engagement", pt.size=1.5, shuffle=T) + DimPlot(cd8, group.by="Experiment")
```


#### 2.1.2 Batch Correction
Based on the results of the previous section, there is a clear experimental batch effect that needs to be corrected.
We will use RunFastMNN function from SeuratWrappers package, which is a fast implementation of the original MNN algorithm.


```{r}
cd8_mnn = RunFastMNN(object.list=SplitObject(cd8, split.by="Experiment"), features=3000, verbose=F)

# UMAP embedding & Clusting
cd8_mnn = RunUMAP(cd8_mnn, dims=1:15, min.dist=0.3, spread=2.5, reduction="mnn", verbose=F) %>% FindNeighbors(dims=1:15, reduction="mnn", k.param=40, verbose=F) %>% FindClusters(resolution=1.5, verbose=F)
options(repr.plot.width=6, repr.plot.height=5)
DimPlot(cd8_mnn, label=TRUE, pt.size=1.5)
```

```{r}
cd8_mnn$Engagement = factor(cd8_mnn$Engagement, levels=c("medium-exposed", "never-engaged", "engager", "super-engaged"))

# Plot Cells
options(repr.plot.width=12, repr.plot.height=5)
DimPlot(cd8_mnn, pt.size=1.5, shuffle=T, group.by="Experiment") + DimPlot(cd8_mnn, pt.size=1.5, group.by="Engagement", shuffle=T)
```

#### 2.1.3 Save Seurat Object for Downstream Analysis
From the above results, we have successfully corrected the batch effect and save the cd8_mnn object for downstream analysis.

```{r}
saveRDS(cd8_mnn, "CD8_MNN_Seurat.rds")
```



### 2.2 Analyze CD4+ T cells
#### 2.2.1 Generate Seurat Object and Embed Cells
Just like before, we generate a new Seurat object of only CD4+ T cells and analyze them using the Guided Clustering Tutorial workflow.

```{r}
cd4_counts = GetAssayData(cd4_temp, slot="counts")
cd4 = CreateSeuratObject(counts=cd4_counts, project="cd4", min.cells=0, min.features=0)
cd4$Engagement = factor(cd4_temp$Engagement, levels=c("medium-exposed", "never-engaged", "nonengager", "engager", "super-engaged"))
cd4$Engagement = gsub("nonengager", "never-engaged", cd4$Engagement)
cd4$Organoid = cd4_temp$Organoid
cd4$Experiment = cd4_temp$Experiment
cd4$Plate = cd4_temp$Plate

# Normalized & Scale & PCA dimentionality reduction
cd4 = NormalizeData(cd4, verbose=F) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(features=rownames(cd8), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
cd4 = RunUMAP(cd4, dims=1:12, min.dist=0.2, verbose=F) %>% FindNeighbors(dims=1:12, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd4) + DimPlot(cd4, group.by="Engagement", pt.size=1.5, shuffle=T) + DimPlot(cd4, group.by="Experiment")
```



#### 2.2.2 Further Subset CD4+ T Cells
It appears that there are two subtuypes of T cells, best distinguished by "IL17RB".

```{r}
options(repr.plot.width=6, repr.plot.height=5)
FeaturePlot(cd4, order=T, "IL17RB", label=T)
```

So, we will further subset CD4+ T cells into two distinctive IL17RB+ and IL17RB- subsets.

```{r}
cd4_temp_1 = subset(cd4, idents = c(1, 7, 10))
cd4_temp_1$Engagement = cd4$Engagement; cd4_temp_1$Organoid = cd4$Organoid
cd4_temp_1$Experiment = cd4$Experiment; cd4_temp_1$Plate = cd4$Plate
cd4_temp_2 = subset(cd4, idents = c(1, 7, 10), invert = T)
cd4_temp_2$Engagement = cd4$Engagement; cd4_temp_2$Organoid = cd4$Organoid
cd4_temp_2$Experiment = cd4$Experiment; cd4_temp_2$Plate = cd4$Plate
```



#### 2.2.3 Analyze CD4+/IL17RB- T Cells
#### 2.2.3.1 Generate Seurat Object and Embed Cells
Just like before, we generate a new Seurat object of only CD4+ IL17RB- T cells and analyze them using the Guided Clustering Tutorial workflow.

```{r}
cd4_counts = GetAssayData(cd4_temp_2, slot = "counts")
cd4_2 = CreateSeuratObject(counts = cd4_counts, project = "cd4", min.cells = 0, min.features = 0)
cd4_2$Engagement = factor(cd4_temp_2$Engagement, levels = c("medium-exposed", "never-engaged", "nonengager", "engager", "super-engaged"))
cd4_2$Engagement = gsub("nonengager", "never-engaged", cd4_2$Engagement)
cd4_2$Organoid = cd4_temp_2$Organoid
cd4_2$Experiment = cd4_temp_2$Experiment
cd4_2$Plate = cd4_temp_2$Plate

# Normalized & Scale & PCA dimentionality reduction
cd4_2 = NormalizeData(cd4_2, verbose=F) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(features=rownames(cd8), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
cd4_2 = RunUMAP(cd4_2, dims=1:12, min.dist=0.2, verbose=F) %>% FindNeighbors(dims=1:12, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd4_2) + DimPlot(cd4_2, group.by="Engagement", pt.size=1.5, shuffle=T) + DimPlot(cd4_2, group.by="Experiment")
```

#### 2.2.3.2 Batch Correction
Based on the results of the previous section, there is a clear experimental batch effect that needs to be corrected.

```{r}
# There is a clear batch effect between experiments. Correct for it with MNN
cd4_2_mnn = RunFastMNN(object.list = SplitObject(cd4_2, split.by = "Experiment"), features = 2000, verbose=F)

# UMAP embedding & Clustering
cd4_2_mnn = RunUMAP(cd4_2_mnn, dims=1:15, min.dist=0.3, spread=2.5, reduction="mnn", verbose=F) %>% FindNeighbors(dims=1:15, reduction="mnn", k.param=40, verbose=F) %>% FindClusters(resolution=2, verbose=F)
cd4_2_mnn$Engagement = factor(cd4_2_mnn$Engagement, levels=c("medium-exposed", "never-engaged", "engager", "super-engaged"))

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd4_2_mnn, label=T, pt.size=1.5) + DimPlot(cd4_2_mnn, label=F, pt.size=1.5, group.by="Experiment") +
DimPlot(cd4_2_mnn, label=F, pt.size=1.5, group.by="Engagement", shuffle=T)
```

It is apparent that there is still remaining batch-specific clusters.Since we are mostly interested in shared behavioral mode of actions, we will only keep cluster of cells that are shared between two experiments.

```{r}
cd4_2_mnn_2 = subset(cd4_2_mnn, idents=c(0, 1), invert=T)
counts = GetAssayData(cd4_2_mnn_2, slot = "counts")
cd4_2_2 = CreateSeuratObject(counts = counts, project = "cd4", min.cells = 0, min.features = 0)
cd4_2_2$Engagement = factor(cd4_2_mnn_2$Engagement, levels = c("medium-exposed", "never-engaged", "nonengager", "engager", "super-engaged"))
cd4_2_2$Engagement = gsub("nonengager", "never-engaged", cd4_2_2$Engagement)
cd4_2_2$Organoid = cd4_2_mnn_2$Organoid
cd4_2_2$Experiment = cd4_2_mnn_2$Experiment
cd4_2_2$Plate = cd4_2_mnn_2$Plate

# Normalized & Scale & PCA dimentionality reduction
cd4_2_2 = NormalizeData(cd4_2_2, verbose=F) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(features=rownames(cd8), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
cd4_2_2 = RunUMAP(cd4_2_2, dims=1:12, min.dist=0.2, verbose=F) %>% FindNeighbors(dims=1:12, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd4_2_2) + DimPlot(cd4_2_2, group.by="Engagement", pt.size=1.5, shuffle=T) + DimPlot(cd4_2_2, group.by="Experiment")
```

```{r}
# Correct for the experimental batch effect
# There is a clear batch effect between experiments. Correct for it with MNN
cd4_2_2_mnn = RunFastMNN(object.list = SplitObject(cd4_2_2, split.by="Experiment"), features=3000, verbose=F)

# UMAP embedding & Clusting
cd4_2_2_mnn = RunUMAP(cd4_2_2_mnn, dims=1:15, min.dist=0.2, spread=0.7, reduction="mnn", verbose=F) %>% FindNeighbors(dims=1:15, reduction="mnn", k.param=40, verbose=F) %>% FindClusters(resolution=2.5, verbose=F)
cd4_2_2_mnn$Engagement = factor(cd4_2_2_mnn$Engagement, levels = c("medium-exposed", "never-engaged", "engager", "super-engaged"))

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd4_2_2_mnn, label=T, pt.size=1.5) +
DimPlot(cd4_2_2_mnn, label=F, pt.size=1.5, group.by="Experiment") +
DimPlot(cd4_2_2_mnn, label=F, pt.size=1.5, group.by="Engagement", shuffle=T)
```

#### 2.2.3.3 Save Seurat Object for Downstream Analysis
From the above results, we have successfully corrected the batch effect and save the cd4_2_2_mnn object for downstream analysis.

```{r}
saveRDS(cd4_2_2_mnn, "CD4_IL17RBNeg_MNN_Seurat.rds")
```



#### 2.2.4 Analyze CD4+/IL17RB+ T Cells
#### 2.2.4.1 Generate Seurat Object and Embed Cells
Just like before, we generate a new Seurat object of only CD4+ IL17RB+ T cells and analyze them using the Guided Clustering Tutorial workflow.

```{r}
cd4_counts = GetAssayData(cd4_temp_1, slot="counts")
cd4_3 = CreateSeuratObject(counts = cd4_counts, project = "cd4", min.cells = 0, min.features = 0)
cd4_3$Engagement = factor(cd4_temp_1$Engagement, levels = c("medium-exposed", "never-engaged", "nonengager", "engager", "super-engaged"))
cd4_3$Engagement = gsub("nonengager", "never-engaged", cd4_3$Engagement)
cd4_3$Organoid = cd4_temp_1$Organoid
cd4_3$Experiment = cd4_temp_1$Experiment
cd4_3$Plate = cd4_temp_1$Plate

# Normalized & Scale & PCA dimentionality reduction
cd4_3 = NormalizeData(cd4_3, verbose=F) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(features=rownames(cd8), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
cd4_3 = RunUMAP(cd4_3, dims=1:12, min.dist=0.2, verbose=F) %>% FindNeighbors(dims=1:12, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)
cd4_3$Engagement = factor(cd4_3$Engagement, levels = c("medium-exposed", "never-engaged", "engager", "super-engaged"))

# Plot
options(repr.plot.width=17, repr.plot.height=5)
DimPlot(cd4_3) + DimPlot(cd4_3, group.by="Engagement", pt.size=1.5, shuffle=T) + DimPlot(cd4_3, group.by="Experiment")
```

#### 2.2.4.2 Save Seurat Object for Downstream Analysis
As this CD4+ T cells are only appeared in one experiment series, there is no need to batch correct, and we will save the cd4_3 object for downstream analysis.

```{r}
saveRDS(cd4_3, "CD4_IL17RBPos_MNN_Seurat.rds")
```



## 3. Generate a Seurat Object of all T Cells
### 3.1 Generate the Object and Subset to Have Only Verified Cells
As the final step in this workflow, we will generate a seurat object using only the verified cells (from previous analysis steps) and embed them in one single UMAP.
To this aim, we again read-in the original Seurat object and subset it to have only verified cells.

```{r}
#teg_orig = readRDS("1_set_N2_3_13T_N0_EXTENDED_MNN_umap.rds")
cd8 = readRDS("CD8_MNN_Seurat.rds")
cd4_1 = readRDS("CD4_IL17RBNeg_MNN_Seurat.rds")
cd4_2 = readRDS("CD4_IL17RBPos_MNN_Seurat.rds")
cells = data.frame(Cells = c(colnames(cd8), colnames(cd4_1), colnames(cd4_2)),
                   Subset = c(rep("CD8", ncol(cd8)), rep("CD4 IL17-", ncol(cd4_1)), rep("CD4 IL17+", ncol(cd4_2))))
teg_orig = subset(teg_orig, cells = cells$Cells)
teg_orig$Subset = cells$Subset
```

```{r}
# Generate a new Seurat object
teg_counts = GetAssayData(teg_orig, slot = "counts")
teg = CreateSeuratObject(counts = teg_counts, project = "teg", min.cells = 0, min.features = 0)
teg$Engagement = teg_orig$Engagement
teg$Engagement = gsub("nonengager", "never-engaged", teg$Engagement)
teg$Organoid = teg_orig$Organoid
teg$Experiment = teg_orig$Experiment
teg$Plate = teg_orig$Plate
teg$Subset = teg_orig$Subset
teg[["percent.mt"]] = PercentageFeatureSet(teg, pattern = "MT-")
teg[["percent.rb"]] = PercentageFeatureSet(teg, pattern = "^RP[LS]")
VlnPlot(teg, features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "percent.rb"), ncol = 4)
```

### 3.2 Normalize & Scale & Reduce dimentionality of the data

```{r}
# Remove mito and rb genes
mt.genes = rownames(teg)[grep("^MT-", rownames(teg))]
rb.genes = rownames(teg)[grep("^RP[LS]", rownames(teg))]
genes = c(mt.genes, rb.genes)
genes.keep = setdiff(rownames(teg), genes)
teg = subset(teg, features = genes.keep)
rm(genes, genes.keep, mt.genes, rb.genes)

# Normalized & Scale & PCA dimentionality reduction
teg = NormalizeData(teg, verbose=F) %>% FindVariableFeatures(nfeatures = 3000) %>% ScaleData(features=rownames(cd8), verbose=F) %>% RunPCA(verbose=F)

# UMAP embedding & Clusting
teg = RunUMAP(teg, dims=1:15, min.dist=0.2, verbose=F) %>% FindNeighbors(dims=1:15, k.param=30, verbose=F) %>% FindClusters(resolution=2, verbose=F)

# Plot
options(repr.plot.width=12, repr.plot.height=5)
DimPlot(teg) + DimPlot(teg, group.by = "Experiment")
```

### 3.3 Batch Correction
Based on the results of the previous section, there is a clear experimental batch effect that needs to be corrected.

```{r}
teg_mnn = RunFastMNN(object.list = SplitObject(teg, split.by="Experiment"), features=3000, verbose=F)

# UMAP embedding & Clusting
teg_mnn = RunUMAP(teg_mnn, dims=1:15, min.dist=0.3, spread=1, reduction="mnn", verbose=F) %>% FindNeighbors(dims=1:15, reduction="mnn", k.param=40, verbose=F) %>% FindClusters(resolution=1.5, verbose=F)
teg_mnn$Subset = factor(teg_mnn$Subset, levels = c("CD4 IL17-", "CD4 IL17+", "CD8"))
teg_mnn$Engagement = factor(teg_mnn$Engagement, levels = c("medium-exposed", "never-engaged", "engager", "super-engaged"))

# Plot
options(repr.plot.width=12, repr.plot.height=10)
DimPlot(teg_mnn, label=T, pt.size=1.5) + DimPlot(teg_mnn, label=F, pt.size=1.5, group.by="Experiment", shuffle=T) +
DimPlot(teg_mnn, label=F, pt.size=1.5, group.by="Subset", shuffle=T) +
DimPlot(teg_mnn, label=F, pt.size=1.5, group.by="Engagement", shuffle=T)
```

### 3.4 Save Seurat Object
From the above results, we have successfully corrected the batch effect and save the teg_mnn object.

```{r}
saveRDS(teg_mnn, "All_TEGs_MNN_Seurat.rds")
```

```{r}
sessionInfo()
```

